From 5887a0d0722e08d2a4d168ff21e99e5b0276422e Mon Sep 17 00:00:00 2001
From: John Galt <johngaltfirstrun@gmail.com>
Date: Fri, 5 Jun 2026 11:43:53 -0400
Subject: [PATCH 13/16] surfaceflinger: arr: handle no vote the same as idle

If applicable, otherwise fallback to old behavior.

Utilizes idle rate significantly more by reducing intermediate 60hz (due
to min in normal group).
---
 .../Scheduler/RefreshRateSelector.cpp           | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp b/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp
index 28a23c1..a8f6ebc 100644
--- a/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp
+++ b/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp
@@ -817,7 +817,22 @@ auto RefreshRateSelector::getRankedFrameRatesLocked(const std::vector<LayerRequi
 
     if (layers.empty() || noVoteLayers == layers.size()) {
         ALOGV("No layers with votes");
-        const auto ranking = rankFrameRates(anchorGroup, RefreshRateOrder::Descending);
+        auto noVoteBestMode = mPrimaryFrameRates.end();
+        for (auto it = mPrimaryFrameRates.begin(); it != mPrimaryFrameRates.end(); ++it) {
+            if (it->modePtr->getGroup() != anchorGroup) continue;
+            if (isStrictlyLess(it->fps, kMinSupportedFrameRate)) continue;
+            if (noVoteBestMode == mPrimaryFrameRates.end() ||
+                isStrictlyLess(it->fps, noVoteBestMode->fps)) {
+                noVoteBestMode = it;
+            }
+        }
+        FrameRateRanking ranking;
+        if (noVoteBestMode != mPrimaryFrameRates.end()) {
+            constexpr float kMaxScore = std::numeric_limits<float>::max();
+            ranking.push_back(ScoredFrameRate{*noVoteBestMode, kMaxScore});
+        } else {
+            ranking = rankFrameRates(anchorGroup, RefreshRateOrder::Descending);
+        }
         SFTRACE_FORMAT_INSTANT("%s (No layers with votes)",
                                to_string(ranking.front().frameRateMode.fps).c_str());
         return {ranking, kNoSignals};
-- 
2.47.3

